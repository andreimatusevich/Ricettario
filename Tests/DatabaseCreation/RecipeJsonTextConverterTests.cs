using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using ApprovalTests;
using ApprovalTests.Reporters;
using MoreLinq;
using Newtonsoft.Json;
using NUnit.Framework;
using Ricettario.Shared.Model;

namespace Tests.DatabaseCreation
{
    public static class Extensions
    {
        public static string TrimEx(this string text)
        {
            return text?.Trim(' ', ';', ',', ':');
        }

        public static bool IsAlike(this string text, params string[] ethalons)
        {
            var normalized = text?.TrimEx().ToLower();
            return ethalons.Select(e => e?.TrimEx().ToLower()).Any(e => e == normalized);
        }
    }

    public class TextRecipeParser
    {
        private readonly IFactory<Recipe> _factory;

        private static Dictionary<string, Action<Recipe, IEnumerable<string>>> _rules = new Dictionary
            <string, Action<Recipe, IEnumerable<string>>>
        {
            {
                "Name", (recepie, lines) => recepie.Name = lines.First().Trim()
            },
            {
                "RecipeYield",
                (recepie, lines) => recepie.RecipeYield = String.Join(" ", lines.Skip(1).TakeWhile(l => !String.IsNullOrEmpty(l)))
            },
            {
                "Ingredients",
                (recepie, lines) =>
                    recepie.RecipeIngredients =
                        lines.SkipWhile(l => !l.IsAlike("Ingredients"))
                            .Skip(1)
                            .SkipWhile(l => String.IsNullOrEmpty(l))
                            .TakeWhile(l => !String.IsNullOrEmpty(l)).ToList()
            },
            {
                "Instructions",
                (recepie, lines) =>
                    recepie.RecipeInstructions =
                        lines.SkipWhile(l => !l.IsAlike("Method", "Directions"))
                            .Skip(1).Where(l => !String.IsNullOrEmpty(l)).ToList()
            }
        };

        public TextRecipeParser(IFactory<Recipe> factory)
        {
            _factory = factory;
        }

        public Recipe FromLines(IEnumerable<string> lines)
        {
            var normalizedLines = lines.SkipWhile(l => String.IsNullOrEmpty(l.TrimEx())).Select(l => l.Trim()).ToList();
            var recepie = _factory.Create();
            _rules.Values.ForEach(r => r(recepie, normalizedLines));
            return recepie;
        }
    }

    public class RecipeJsonTextConverterTests
    {
        [Test]
        [UseReporter(typeof(WinMergeReporter))]
        public void Simple()
        {
            var fileName = @"..\..\DatabaseCreation\Data\simple_recepie.txt";
            TestParse(fileName);
        }

        [Test]
        [UseReporter(typeof(WinMergeReporter))]
        public void GoogleDocs()
        {
            var fileName = @"..\..\DatabaseCreation\Data\google_docs_recepie.txt";
            TestParse(fileName);
        }

        private static void TestParse(string fileName)
        {
            var lines = File.ReadAllLines(fileName);

            var parser = new TextRecipeParser(new TestRecipeFactory());
            var recepie = parser.FromLines(lines);

            var result = JsonConvert.SerializeObject(recepie, Formatting.Indented,
                new JsonSerializerSettings {NullValueHandling = NullValueHandling.Ignore});
            Approvals.Verify(result);
        }
    }


    public class TestRecipeFactory: IFactory<Recipe>
    {
        public Recipe Create()
        {
            return new Recipe()
            {
            };
        }
    }
}