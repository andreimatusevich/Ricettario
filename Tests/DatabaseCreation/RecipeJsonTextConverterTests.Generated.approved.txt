{
  "Version": "0001-01-01T00:00:00",
  "Name": "Classic Buttermilk Waffles",
  "RecipeIngredients": [
    "2 large eggs",
    "1 3/4 cups buttermilk",
    "1/2 cup (1 stick) butter, melted and cooled to room temperature",
    "2 teaspoons vanilla",
    "2 cups Round Table Unbleached Pastry OR 1 3/4 cups King Arthur Unbleached All-Purpose Flour",
    "2 tablespoons sugar (leave this out if you want savory waffles, such as the type you'd use as a base for creamed chipped beef or fried chicken)",
    "2 teaspoons baking powder",
    "1 teaspoon soda",
    "1 teaspoon salt",
    "1/2 cup pecan meal or almond flour (optional)"
  ],
  "RecipeInstructions": [
    "In a medium-sized mixing bowl, beat together the eggs, buttermilk, melted butter or margarine and vanilla. In another bowl whisk together the dry ingredients. Combine the wet and dry ingredients just until almost smooth.",
    "Spray your waffle iron with a non-stick cooking spray before preheating it. For an 8-inch round waffle iron, use about 1/3 cup batter. Cook for 2 to 3 minutes, until the iron stops steaming. Yield: about ten 8-inch waffles."
  ],
  "RecipeDescription": "This recipe is for a very typical plain waffle, crisp and golden, perfect with maple syrup and butter or berries and whipped cream. When made with pastry flour it will be extra crispy, and light as air inside. When made with all-purpose flour the waffle has a bit more body -- still light, but chewier on the inside. Waffles are best consumed as soon as they're baked, but in a pinch you may place them on a wire rack to cool, wrap tightly to store in the refrigerator, then reheat for 6 minutes in a 350°F oven. The optional pecan meal or almond flour add a nice nutty flavor."
}
-------------------
Name:
	Classic Buttermilk Waffles

Yield:

Ingredients:
	2 large eggs
	1 3/4 cups buttermilk
	1/2 cup (1 stick) butter, melted and cooled to room temperature
	2 teaspoons vanilla
	2 cups Round Table Unbleached Pastry OR 1 3/4 cups King Arthur Unbleached All-Purpose Flour
	2 tablespoons sugar (leave this out if you want savory waffles, such as the type you'd use as a base for creamed chipped beef or fried chicken)
	2 teaspoons baking powder
	1 teaspoon soda
	1 teaspoon salt
	1/2 cup pecan meal or almond flour (optional)

Directions:
	In a medium-sized mixing bowl, beat together the eggs, buttermilk, melted butter or margarine and vanilla. In another bowl whisk together the dry ingredients. Combine the wet and dry ingredients just until almost smooth.
	Spray your waffle iron with a non-stick cooking spray before preheating it. For an 8-inch round waffle iron, use about 1/3 cup batter. Cook for 2 to 3 minutes, until the iron stops steaming. Yield: about ten 8-inch waffles.

Description:
	This recipe is for a very typical plain waffle, crisp and golden, perfect with maple syrup and butter or berries and whipped cream. When made with pastry flour it will be extra crispy, and light as air inside. When made with all-purpose flour the waffle has a bit more body -- still light, but chewier on the inside. Waffles are best consumed as soon as they're baked, but in a pinch you may place them on a wire rack to cool, wrap tightly to store in the refrigerator, then reheat for 6 minutes in a 350°F oven. The optional pecan meal or almond flour add a nice nutty flavor.

Author:

Tags: